core/include/cpuid.h

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2018 Alexandro Sanchez Bach <alexandro@phi.nz>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *   1. Redistributions of source code must retain the above copyright notice,
+ *      this list of conditions and the following disclaimer.
+ *
+ *   2. Redistributions in binary form must reproduce the above copyright
+ *      notice, this list of conditions and the following disclaimer in the
+ *      documentation and/or other materials provided with the distribution.
+ *
+ *   3. Neither the name of the copyright holder nor the names of its
+ *      contributors may be used to endorse or promote products derived from
+ *      this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef HAX_CORE_CPUID_H_
+#define HAX_CORE_CPUID_H_
+
+#include "hax_types.h"
+
+#define CPUID_CACHE_SIZE 6
+
+#define CPUID_REG_EAX 0
+#define CPUID_REG_ECX 1
+#define CPUID_REG_EDX 2
+#define CPUID_REG_EBX 3
+
+/*
+ * X86 Features
+ * ============
+ * Each feature is 32-bit integer key encoding the following bitfields:
+ * - index        : 5;  Array index where the cached features reside
+ * - root_key     : 6;  Encodes the EAX register value (see notes)
+ * - leaf_reg     : 2;  Leaf register index
+ * - leaf_key     : 5;  Leaf register value
+ * - value_reg    : 2;  Output register index
+ * - value_bit    : 5;  Bit position of the feature
+ *
+ * Additionally root_key is derived from the EAX register as follows:
+ * - root_key_lo  : 5;  Masked least-significant 5-bits of EAX
+ * - root_key_hi  : 1;  Masked most-significant 1-bit of EAX
+ *
+ * This allows representing any feature obtained via:
+ * - Root register (EAX): 00h-1Fh and 80000000h-8000001Fh
+ * - Leaf register (ECX, EDX, EBX): 00h-1Fh
```

Comment:
Regarding to your points:

1. You are right, I will use the leaf/subleaf naming to remain consistent with the Intel SDM (which is also what I quoted in the code comments). My previous terms were chosen arbitrarily.

2. True, I have transformed that field into boolean 1-bit field named `subleaf_used`.

3. True, I have extended `leaf_hi` (previously `root_key_hi`) to 2-bits. However, I have kept the remaining "key" fields unmodified (5-bits), since no leaf or subleaf has a lower part larger than 0x1F, both on Intel and AMD reference manuals. I think we should only extend fields when necessary to avoid extending the wrong fields and later running out of extra bits for other fields. Since these fields are only consumed **internally** by *cpuid.h*, increasing their size can be done at any time easily.